android/expoview/src/main/java/host/exp/exponent/ExpoUpdatesAppLoader.kt

```diff
@@ -0,0 +1,468 @@
+import expo.modules.updates.loader.EmbeddedLoader
+import expo.modules.updates.loader.FileDownloader
+import expo.modules.updates.loader.LoaderTask
+import expo.modules.updates.loader.LoaderTask.BackgroundUpdateStatus
+import expo.modules.updates.loader.LoaderTask.LoaderTaskCallback
+import expo.modules.updates.manifest.Manifest
+import expo.modules.updates.manifest.ManifestFactory
+import expo.modules.updates.manifest.raw.RawManifest
+import expo.modules.updates.selectionpolicy.LauncherSelectionPolicyFilterAware
+import expo.modules.updates.selectionpolicy.LoaderSelectionPolicyFilterAware
+import expo.modules.updates.selectionpolicy.ReaperSelectionPolicyDevelopmentClient
+import expo.modules.updates.selectionpolicy.SelectionPolicy
+import host.exp.exponent.di.NativeModuleDepsProvider
+import host.exp.exponent.exceptions.ManifestException
+import host.exp.exponent.kernel.ExperienceKey
+import host.exp.exponent.kernel.ExpoViewKernel
+import host.exp.exponent.kernel.Kernel
+import host.exp.exponent.kernel.KernelConfig
+import host.exp.exponent.storage.ExponentSharedPreferences
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.File
+import java.util.*
+import javax.inject.Inject
+
+private const val UPDATE_AVAILABLE_EVENT = "updateAvailable"
+private const val UPDATE_NO_UPDATE_AVAILABLE_EVENT = "noUpdateAvailable"
+private const val UPDATE_ERROR_EVENT = "error"
+
+class ExpoUpdatesAppLoader @JvmOverloads constructor(
+  private val manifestUrl: String,
+  private val callback: AppLoaderCallback,
+  private val useCacheOnly: Boolean = false
+) {
+  @Inject
+  lateinit var exponentManifest: ExponentManifest
+
+  @Inject
+  lateinit var exponentSharedPreferences: ExponentSharedPreferences
+
+  @Inject
+  lateinit var databaseHolder: DatabaseHolder
+
+  @Inject
+  lateinit var kernel: Kernel
+
+  enum class AppLoaderStatus {
+    CHECKING_FOR_UPDATE, DOWNLOADING_NEW_UPDATE
+  }
+
+  var isEmergencyLaunch = false
+    private set
+  var isUpToDate = true
+    private set
+  var status: AppLoaderStatus? = null
+    private set
+  var shouldShowAppLoaderStatus = true
+    private set
+  private var isStarted = false
+
+  interface AppLoaderCallback {
+    fun onOptimisticManifest(optimisticManifest: RawManifest)
+    fun onManifestCompleted(manifest: RawManifest)
+    fun onBundleCompleted(localBundlePath: String?)
+    fun emitEvent(params: JSONObject)
+    fun updateStatus(status: AppLoaderStatus)
+    fun onError(e: Exception)
+  }
+
+  private var _updatesConfiguration: UpdatesConfiguration? = null
+  private var _updatesDirectory: File? = null
+  private var _fileDownloader: FileDownloader? = null
+  private var _selectionPolicy: SelectionPolicy? = null
+  private var _launcher: Launcher? = null
+
+  val updatesConfiguration: UpdatesConfiguration
+    get() {
+      return checkNotNull(_updatesConfiguration) { "Tried to access UpdatesConfiguration before it was set" }
+    }
```

Comment:
The compiler will suggest removing `checkNotNull` since it knows/thinks that field is non-null. 

What I really want is `lateinit` with a custom error message, but unfortunately that isn't supported:
```kotlin
lateinit var updatesConfiguration: UpdatesConfiguration
    private set
```

Unless we're okay with a lateinit error (non-custom error) then this backing field is the best we can do.

Reply:
I think the lateinit error is fine. If this code were written from scratch, we'd probably use lateinit.